File: app/src/main/java/com/veganbeauty/app/data/local/dao/UserProductExpiryDao.java
package com.veganbeauty.app.data.local.dao;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;

import com.veganbeauty.app.data.local.entities.UserProductExpiryEntity;

import java.util.List;
import kotlinx.coroutines.flow.Flow;

@Dao
public interface UserProductExpiryDao {
    @Query("SELECT * FROM user_product_expiry WHERE userId = :userId")
    Flow<List<UserProductExpiryEntity>> getProductsByUserIdFlow(String userId);

    @Query("SELECT * FROM user_product_expiry WHERE userId = :userId")
    List<UserProductExpiryEntity> getProductsByUserId(String userId);

    @androidx.annotation.Nullable
    @Query("SELECT * FROM user_product_expiry WHERE userId = :userId AND productId = :productId LIMIT 1")
    UserProductExpiryEntity getProductById(String userId, String productId);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    long[] insertUserProducts(List<UserProductExpiryEntity> products);

    @Query("SELECT COUNT(*) FROM user_product_expiry WHERE userId = :userId")
    int getProductCountByUserId(String userId);

    @Query("DELETE FROM user_product_expiry WHERE userId = :userId AND productId = :productId")
    int deleteUserProductExpiry(String userId, String productId);
}

File: app/src/main/java/com/veganbeauty/app/data/local/dao/UserProductExpiryDao.kt
package com.veganbeauty.app.data.local.dao

import androidx.room.*
import com.veganbeauty.app.data.local.entities.UserProductExpiryEntity
import kotlinx.coroutines.flow.Flow

@Dao
interface UserProductExpiryDao {
    @Query("SELECT * FROM user_product_expiry WHERE userId = :userId")
    fun getProductsByUserIdFlow(userId: String): Flow<List<UserProductExpiryEntity>>

    @Query("SELECT * FROM user_product_expiry WHERE userId = :userId")
    suspend fun getProductsByUserId(userId: String): List<UserProductExpiryEntity>

    @Query("SELECT * FROM user_product_expiry WHERE userId = :userId AND productId = :productId LIMIT 1")
    suspend fun getProductById(userId: String, productId: String): UserProductExpiryEntity?

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertUserProducts(products: List<UserProductExpiryEntity>)

    @Query("SELECT COUNT(*) FROM user_product_expiry WHERE userId = :userId")
    suspend fun getProductCountByUserId(userId: String): Int

    @Query("DELETE FROM user_product_expiry WHERE userId = :userId AND productId = :productId")
    suspend fun deleteUserProductExpiry(userId: String, productId: String)
}
